R/transform.R

```diff
@@ -31,167 +36,239 @@
-      cwt <- cwt[ , , 1, ] * torch_complex(1, 0) + cwt[ , , 2, ] * torch_complex(0, 1)
-    } else {
-    # Remove the chn_out dimension to obtain an object of shape [batch_size,n_scales,signal_length].
-      cwt = cwt$squeeze(3)
-    }
-    # Remove batch dimension if single example
-    if (num_examples == 1) {
-      cwt <- cwt$squeeze(1)
-    }
-    cwt
-  },
-  # Determines the optimal scale distribution (see Torrence & Compo, Eq. 9-10).
-  compute_optimal_scales = function() {
-    J <- floor((1 / self$dj) * log2(self$signal_length * self$dt / self$scale_minimum))
-    scales <- self$scale_minimum * 2^(self$dj * torch_arange(0, J))
-    scales
-  },
-  # Performs CWT and converts to a power spectrum (scalogram). See Torrence & Compo, Section 4d.
-  # Expects a batch of input signals of shape `[batch_size,signal_length]` and
-  # returns a scalogram for each signal `[batch_size,n_scales,signal_length]`
-  # unbias: whether to unbias the power spectrum, as in Liu et al. 2007.
-  power = function(x, unbias = FALSE) {
-    if (unbias) (torch_abs((self$forward(x))$T)^2 / self$scales)$T else torch_abs(self$forward(x))^2
-  },
-  get_padding = function(padding_type, kernel_size) {
-    if (padding_type == "same") floor((kernel_size - 1) / 2) else 0
-  },
-  # Choose s0 so that the equivalent Fourier period is 2 * dt. See Torrence & Compo Sections 3f and 3h.
-  compute_minimum_scale = function() {
-    dt <- self$dt
-    f <- self$wavelet$fourier_period
-    func_to_solve <- function(s) {
-      f(s) - 2 * dt
-    }
-    uniroot(func_to_solve, c(0,10))$root
-  },
-  fourier_period = function(s) {
-    self$wavelet$fourier_period(s)
-  },
-  scale_from_period = function(p) {
-    self$wavelet$scale_from_period(p)
-  },
-  fourier_periods = function() {
-    self$fourier_period(self$scales)
-  },
-  fourier_frequencies = function() {
-    torch_reciprocal(self$fourier_periods())
-  },
-  # returns a pair of tensors that describes the cone of influence as a curve of time and scales
-  coi = function(t_min, t_max) {
-    w_coi <- self$wavelet$coi(self$scales)
-    t_mid <- t_min + (t_max - t_min) / 2
+  "WaveletTransform",
+  lock_objects = FALSE,
+  public = list(
+    #' @field filters the filters used in the time-domain implementation
+    filters = NULL,
+    #' @field is_complex_wavelet whether the wavelet has complex values
+    is_complex_wavelet = NULL,
+    #' @field convs the `nn_conv1d()` modules used in the time-domain implementation
+    convs = NULL,
+    #' @description initialize the main object, and possibly pre-compute required fields.
+    #' @param signal_length length of the signal to be processed
+    #' @param dt sample spacing, default is 1
+    #' @param dj scale distribution parameter, default is 0.125
+    #' @param wavelet wavelet object
+    #' @param fourier whether to run in the frequency domain
+    initialize = function(signal_length, dt = 1, dj = 0.125, wavelet = torchwavelets::Morlet$new(), fourier = TRUE) {
```

Comment:
```suggestion
    initialize = function(signal_length, dt = 1, dj = 0.125, wavelet = NULL, fourier = TRUE) {
```

I'd use the default of NULL and then a `if (is.null(wavelet)) wavelet <- torchwavelets::Morlet$new()`. This kind of self reference can cause some unexpected errors.